waku/v2/discv5/discover.go

```diff
@@ -0,0 +1,357 @@
+	if err != nil {
+		return nil, err
+	}
+
+	discV5 := new(DiscoveryV5)
+	discV5.host = host
+	discV5.params = params
+	discV5.peerCache.rng = rand.New(rand.NewSource(rand.Int63()))
+	discV5.peerCache.recs = make(map[peer.ID]peerRecord)
+	discV5.localnode = localnode
+	discV5.config = discover.Config{
+		PrivateKey: priv,
+		Bootnodes:  params.bootnodes,
+	}
+	discV5.udpAddr = &net.UDPAddr{
+		IP:   ipAddr,
+		Port: params.udpPort,
+	}
+
+	return discV5, nil
+}
+
+func newLocalnode(priv *ecdsa.PrivateKey, ipAddr net.IP, udpPort int, tcpPort int, wakuFlags WakuEnrBitfield, advertiseAddr *net.IP) (*enode.LocalNode, error) {
+	db, err := enode.OpenDB("")
+	if err != nil {
+		return nil, err
+	}
+
+	localnode := enode.NewLocalNode(db, priv)
+	localnode.SetFallbackIP(net.IP{127, 0, 0, 1})
+	localnode.SetFallbackUDP(udpPort)
+
+	if advertiseAddr != nil {
+		localnode.SetStaticIP(*advertiseAddr)
+	}
+
+	localnode.Set(enr.WithEntry(WakuENRField, wakuFlags))
+
+	localnode.Set(enr.IP(ipAddr))
+	localnode.Set(enr.UDP(udpPort))
+	localnode.Set(enr.TCP(tcpPort))
+
+	return localnode, nil
+}
+
+func (d *DiscoveryV5) Start() error {
+	conn, err := net.ListenUDP("udp", d.udpAddr)
+	if err != nil {
+		return err
+	}
+
+	listener, err := discover.ListenV5(conn, d.localnode, d.config)
+	if err != nil {
+		return err
+	}
+
+	d.listener = listener
+
+	log.Info("Started Discovery V5 at %s:%d", d.udpAddr.IP, d.udpAddr.Port)
+
+	return nil
+}
+
+func (d *DiscoveryV5) Stop() {
+	d.listener.Close()
+}
+
+func isWakuNode(node *enode.Node) bool {
+	enrField := new(WakuEnrBitfield)
+	if err := node.Record().Load(enr.WithEntry(WakuENRField, &enrField)); err != nil {
+		if !enr.IsNotFound(err) {
+			log.Error("could not retrieve port for enr ", node)
+		}
+		return false
+	}
+
+	if enrField != nil {
+		return *enrField != uint8(0)
+	}
+
+	return false
+}
+
+func (d *DiscoveryV5) evaluateNode(node *enode.Node) bool {
+	if node == nil || node.IP() == nil {
+		return false
+	}
+
+	enrTCP := new(enr.TCP)
+	if err := node.Record().Load(enr.WithEntry(enrTCP.ENRKey(), enrTCP)); err != nil {
+		if !enr.IsNotFound(err) {
+			log.Error("could not retrieve port for enr ", node)
+		}
+		return false
+	}
+
+	if !isWakuNode(node) {
+		return false
+	}
+
+	address, err := utils.EnodeToMultiAddr(node)
+	if err != nil {
+		log.Error("could not convert enode to multiaddress ", err)
+		return false
+	}
+
+	info, err := peer.AddrInfoFromP2pAddr(address)
+	if err != nil {
+		log.Error("could not obtain peer info from multiaddress ", err)
+		return false
+	}
+
+	if d.host.Network().Connectedness(info.ID) == network.Connected {
+		return false
+	}
+
+	return true
+}
+
+func (c *DiscoveryV5) Advertise(ctx context.Context, ns string, opts ...discovery.Option) (time.Duration, error) {
+	// Get options
+	var options discovery.Options
+	err := options.Apply(opts...)
+	if err != nil {
+		return 0, err
+	}
+
+	// TODO: once discv5 spec introduces capability and topic discovery, implement this function
```

Comment:
this looks like it does nothing yet right?

Reply:
Yeah, this part has not been specified yet.